src/frontend/contexts/products-context.js

```diff
@@ -0,0 +1,66 @@
+import { createContext, useContext, useReducer } from "react";
+import { productsReducer } from "../reducers/productsReducer";
+import {
+  sortByPrice,
+  filterByCategory,
+  filterByRating,
+  filterByPrice,
+} from "../utils";
+import { useData } from "./data-context";
+
+const ProductsContext = createContext();
+
+export const useProducts = () => useContext(ProductsContext);
+
+const initialState = {
+  sortBy: "",
+  category: {
+    CONTEMPORARY_FICTION: false,
+    SELF_HELP: false,
+    BIOGRAPHIES_AUTOBIOGRAPHIES: false,
+    SPIRITUALITY: false,
+    MYTHOLOGY: false,
+    ALL_CATEGORY: false,
+  },
+  collections: {
+    BEST_SELLERS: false,
+    NEW_RELEASES: false,
+    EXPERT_PICKS: false,
+  },
+  filterBy: "",
+  filterPrice: 500,
+};
+
+export function ProductsProvider({ children }) {
+  const { data } = useData();
+
+  const [state, dispatch] = useReducer(productsReducer, initialState);
+
+  const sortedDataByPrice = sortByPrice(data, state.sortBy);
+
+  const filteredDataByCategory = filterByCategory(
+    sortedDataByPrice,
+    state.category.CONTEMPORARY_FICTION,
+    state.category.SELF_HELP,
+    state.category.BIOGRAPHIES_AUTOBIOGRAPHIES,
+    state.category.SPIRITUALITY,
+    state.category.MYTHOLOGY,
+    state.category.ALL_CATEGORY,
+    state.collections.BEST_SELLERS,
+    state.collections.NEW_RELEASES,
+    state.collections.EXPERT_PICKS
+  );
```

Comment:
Suggestion: Instead of passing each category separately, you can pass the entire `category` array in the arguments and then destructure it in the `filterByCategory` function definition